zcash_client_sqlite/src/wallet/init/migrations/shardtree_support.rs

```diff
@@ -149,25 +151,40 @@ impl RusqliteMigration for Migration {
                 stmt_update_block_sapling_tree_size
                     .execute(params![block_end_tree.size(), block_height])?;
 
-                if let Some(nonempty_frontier) = block_end_tree.to_frontier().value() {
-                    trace!(
-                        height = block_height,
-                        frontier = ?nonempty_frontier,
-                        "Inserting frontier nodes",
-                    );
-                    shard_tree
-                        .insert_frontier_nodes(
-                            nonempty_frontier.clone(),
-                            Retention::Checkpoint {
-                                id: BlockHeight::from(block_height),
-                                is_marked: false,
-                            },
-                        )
-                        .map_err(|e| match e {
-                            ShardTreeError::Query(e) => ShardTreeError::Query(e),
-                            ShardTreeError::Insert(e) => ShardTreeError::Insert(e),
-                            ShardTreeError::Storage(_) => unreachable!(),
-                        })?
+                // We only need to load frontiers into the ShardTree that are close enough
+                // to the wallet's known chain tip to fill `PRUNING_DEPTH` checkpoints, so
+                // that ShardTree's witness generation will be able to correctly handle
+                // anchor depths. Loading frontiers further back than this doesn't add any
+                // useful nodes to the ShardTree (as we don't support rollbacks beyond
+                // `PRUNING_DEPTH`, and we won't be finding notes in earlier blocks), and
+                // hurts performance (as frontier importing has a significant Merkle tree
+                // hashing cost).
+                if let Some((nonempty_frontier, (_, latest_height))) = block_end_tree
+                    .to_frontier()
+                    .value()
+                    .zip(block_height_extrema)
+                {
+                    let block_height = BlockHeight::from(block_height);
+                    if block_height + PRUNING_DEPTH >= latest_height {
+                        trace!(
+                            height = u32::from(block_height),
```

Comment:
We already had `block_height` as a `u32` above before it was shadowed. So we can avoid this reconversion by choosing a different name (e.g. `block_height_u32` for the `u32` version.

Reply:
Eh, I'd prefer that we use `BlockHeight` in more places. I added the shadowing here because I wanted it for the addition and comparison, but we need to unwrap for `tracing` macros. The unwrapping cost is literally zero, the compiler will eliminate it entirely as `BlockHeight` is a newtype wrapper around `u32`.